crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl

```diff
@@ -0,0 +1,195 @@
+        let instance_uniform = meshlet_instance_uniforms[instance_id];
+        let world_from_local = affine3_to_square(instance_uniform.world_from_local);
+        let world_position = mesh_position_local_to_world(world_from_local, vec4(vertex.position, 1.0));
+        var clip_position = view.clip_from_world * vec4(world_position.xyz, 1.0);
+        var ndc_position = clip_position.xyz / clip_position.w;
+#ifdef DEPTH_CLAMP_ORTHO
+        ndc_position.z = 1.0 / clip_position.z;
+#endif
+        let viewport_position_xy = ndc_to_uv(ndc_position.xy) * view.viewport.zw;
+
+        // Write vertex to workgroup shared memory
+        viewport_vertices[vertex_id] = vec3(viewport_position_xy, ndc_position.z);
+    }
+
+    workgroupBarrier();
+
+    // Load 1 triangle's worth of vertex data per thread
+    let triangle_id = local_invocation_index;
+    if triangle_id >= meshlet.triangle_count { return; }
+    let index_ids = meshlet.start_index_id + (triangle_id * 3u) + vec3(0u, 1u, 2u);
+    let vertex_ids = vec3(get_meshlet_index(index_ids[0]), get_meshlet_index(index_ids[1]), get_meshlet_index(index_ids[2]));
+    let vertex_0 = viewport_vertices[vertex_ids[2]];
+    let vertex_1 = viewport_vertices[vertex_ids[1]];
+    let vertex_2 = viewport_vertices[vertex_ids[0]];
+    let packed_ids = (cluster_id << 6u) | triangle_id;
+
+    // Compute triangle bounding box
+    let min_x = u32(min3(vertex_0.x, vertex_1.x, vertex_2.x));
+    let min_y = u32(min3(vertex_0.y, vertex_1.y, vertex_2.y));
+    var max_x = u32(ceil(max3(vertex_0.x, vertex_1.x, vertex_2.x)));
+    var max_y = u32(ceil(max3(vertex_0.y, vertex_1.y, vertex_2.y)));
+    max_x = min(max_x, u32(view.viewport.z) - 1u);
+    max_y = min(max_y, u32(view.viewport.w) - 1u);
+    if any(vec2(min_x, min_y) > vec2(max_x, max_y)) { return; }
+
+    // Setup triangle gradients
+    let w_x = vec3(vertex_1.y - vertex_2.y, vertex_2.y - vertex_0.y, vertex_0.y - vertex_1.y);
+    let w_y = vec3(vertex_2.x - vertex_1.x, vertex_0.x - vertex_2.x, vertex_1.x - vertex_0.x);
+    let triangle_double_area = edge_function(vertex_0.xy, vertex_1.xy, vertex_2.xy); // TODO: Reuse earlier calculations and take advantage of summing to 1
+    let vertices_z = vec3(vertex_0.z, vertex_1.z, vertex_2.z) / triangle_double_area;
+    let z_x = dot(vertices_z, w_x);
+    let z_y = dot(vertices_z, w_y);
+
+    // Setup initial triangle equations
+    let starting_pixel = vec2(f32(min_x), f32(min_y)) + 0.5;
+    var w_row = vec3(
+        // TODO: Reuse earlier calculations and take advantage of summing to 1
+        edge_function(vertex_1.xy, vertex_2.xy, starting_pixel),
+        edge_function(vertex_2.xy, vertex_0.xy, starting_pixel),
+        edge_function(vertex_0.xy, vertex_1.xy, starting_pixel),
+    );
+    var z_row = dot(vertices_z, w_row);
+    let view_width = u32(view.viewport.z);
+    var frag_coord_1d_row = min_y * view_width;
+
+    // Rasterize triangle
+    if subgroupAny(max_x - min_x > 4u) {
```

Comment:
what about max_y - min_y being huge? maybe its better to use triangle_double_area?

Reply:
Not really worth doing the whole setup for the scanline calcs in both dimensions. Only worth doing for the inner loop